### PR TITLE
pkgs/formats: add generator for PHP config files

### PR DESCRIPTION
**Pull request by thenhnn**
## Description of changes

Some PHP apps are configurable through evaluation of `config.php` file (file should contain php code that should set options on hashmap), for example [Filesender](https://github.com/filesender/filesender) or [Flarum](https://docs.flarum.org/config/).

This PR adds generation of PHP configs through Nix DSL.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/pkgs-lib/formats.nix
Comment (Ma27) [inline pkgs/pkgs-lib/formats.nix:290]: Why a double quote? This means that `${foo}`, `{$foo}` and `$foo` are substitutable.

Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix:290]: Oops, I didn't know that. Thanks, I will fix.

File: pkgs/pkgs-lib/formats.nix
Comment (drupol) [inline pkgs/pkgs-lib/formats.nix:334]: Could we add `declare(strict_types=1);` ?

**Review by drupol — COMMENTED** [review]
I quickly reviewed and left a comment.

I'm wondering if we could validate the config with `php -l` by default and eventually format it using PSR-12 with PHP-CS-Fixer.

File: pkgs/pkgs-lib/formats.nix
Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix:334]: What's the win from that? 

> Strict typing applies to function calls made from within the file with strict typing enabled, not to the functions declared within that file. If a file without strict typing enabled makes a call to a function that was defined in a file with strict typing, the caller's preference (coercive typing) will be respected, and the value will be coerced.

This only applies for function calls, which is very rare case in configs.

**Comment by thenhnn** [comment]
> I quickly reviewed and left a comment.
> 
> I'm wondering if we could validate the config with `php -l` by default and eventually format it using PSR-12 with PHP-CS-Fixer.

I'm afraid that it will slow down evaluation a lot.

File: pkgs/pkgs-lib/formats.nix
Comment (drupol) [reply pkgs/pkgs-lib/formats.nix:334]: That's true too.

Essentially, I use `declare(strict_types=1);` by default and for promoting best practices when one open a PHP file. To me it should be by default.

Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix:334]: Is it supported only in PHP 8? I can't find mention of versions on PHP website.

Comment (drupol) [reply pkgs/pkgs-lib/formats.nix:334]: It's available since ... PHP 4: https://3v4l.org/IePpo#v4.3.11

Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix]: Done

File: pkgs/pkgs-lib/tests/formats.nix
Comment (drupol) [inline pkgs/pkgs-lib/tests/formats.nix:439]: If adding `attrs.bar = "bar";` just under this line, will it convert it correctly into a single `attrs => [ 'bar' => 'bar', 'foo' => null ]` ?

Comment (thenhnn) [reply pkgs/pkgs-lib/tests/formats.nix:439]: Yeah. From nix perspective `attrs.bar = "bar"; attrs.foo = null;` is equal to `attrs = { bar = "bar"; foo = null; };`.

Comment (drupol) [reply pkgs/pkgs-lib/tests/formats.nix:439]: That's cool then.

**Review by drupol — APPROVED** [review]
LGTM, I'll wait for @jtojnar to give its feedback on this before merging.

File: pkgs/pkgs-lib/formats.nix
Comment (jtojnar) [inline pkgs/pkgs-lib/formats.nix:297]: For easier debugging, it might be nice to format the output so that it  is not squashed into a single unreadable line.

File: pkgs/pkgs-lib/formats.nix
Comment (jtojnar) [inline pkgs/pkgs-lib/formats.nix:336]: `isNull` is deprecated. Use `== null` instead.

File: pkgs/pkgs-lib/formats.nix
Comment (jtojnar) [inline pkgs/pkgs-lib/formats.nix:332]: `name` is unused.

**Review by jtojnar — CHANGES_REQUESTED** [review]

File: nixos/doc/manual/development/settings-options.section.md
Comment (thenhnn) [reply nixos/doc/manual/development/settings-options.section.md:167]: They're used in [SimpleSAMLphp](https://simplesamlphp.org/docs/stable/simplesamlphp-sp.html).
Basically main point of this PR is to supported mixed lists :)

File: pkgs/pkgs-lib/formats.nix
Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix]: Which formatter should I use?

File: pkgs/pkgs-lib/formats.nix
Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix]: Is there any Nix primitive type that is not listed here?

Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix]: I don't know how to make it prettier.

File: nixos/doc/manual/development/settings-options.section.md
Comment (jtojnar) [reply nixos/doc/manual/development/settings-options.section.md:149]: Just add the anchor to the end of the line.

**Comment by jtojnar** [comment]
> > I'm wondering if we could validate the config with `php -l` by default and eventually format it using PSR-12  with PHP-CS-Fixer.
> 
> I'm afraid that it will slow down evaluation a lot.

It would be done at build time so there should be negligible effect on evaluation time.

File: nixos/doc/manual/development/settings-options.section.md
Comment (jtojnar) [reply nixos/doc/manual/development/settings-options.section.md:167]: Maybe specify the order at least?

Comment (thenhnn) [reply nixos/doc/manual/development/settings-options.section.md:167]: What do you mean? The order is `array(<list items>, <keys> => <values>)`.

Comment (jtojnar) [reply nixos/doc/manual/development/settings-options.section.md:167]: Yes, it is not clear from the docs and as demonstrated above, it matters.

File: pkgs/pkgs-lib/formats.nix
Comment (jtojnar) [reply pkgs/pkgs-lib/formats.nix]: Now I wonder if it maybe should be moved into separate file like the java-properties.

File: pkgs/pkgs-lib/formats.nix
Comment (jtojnar) [reply pkgs/pkgs-lib/formats.nix]: php-cs-fixer should be fine. Or alternately, you could do a minimal formatting for legibility by hand by tracking indentation level of the printed value.

Comment (drupol) [reply pkgs/pkgs-lib/formats.nix]: PHP-CS-Fixer is the way to go indeed.

File: pkgs/pkgs-lib/formats.nix
Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix]: IMO it's not big enough for external file.

File: pkgs/pkgs-lib/formats.nix
Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix]: Looks like I'm dumb, but I can't figure out how to make php-cs-fixer to actually format file.

php-cs-fixer fix doesn't work...

Comment (drupol) [reply pkgs/pkgs-lib/formats.nix]: That works here:

```
❯ php-cs-fixer fix --rules=@PSR12 .
PHP CS Fixer 3.51.0 Insomnia by Fabien Potencier, Dariusz Ruminski and contributors.
PHP runtime: 8.2.19
Loaded config default.
 77/77 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

   1) tests/unit/TypedIterableAggregateTest.php
   2) tests/unit/SortIterableAggregateTest.php
   3) tests/static-analysis/closure-iterator.php
   4) src/SortIterableAggregate.php
   5) src/RandomIterableAggregate.php
   6) src/ConcatIterableAggregate.php
   7) src/UnpackIterableAggregate.php
   8) src/FilterIterableAggregate.php
   9) src/NormalizeIterableAggregate.php
  10) src/ChunkIterableAggregate.php
  11) src/PackIterableAggregate.php
  12) src/InterruptableIterableIteratorAggregate.php
  13) src/MapIterableAggregate.php
  14) src/StringIteratorAggregate.php
  15) src/IterableIteratorAggregate.php
  16) src/RandomIntegerAggregate.php
  17) src/ReductionIterableAggregate.php
  18) src/MultipleIterableAggregate.php
  19) src/ReduceIterableAggregate.php
  20) src/LimitIterableAggregate.php
  21) stable-sort.php

Fixed 21 of 77 files in 0.387 seconds, 16.000 MB memory used
```

**Comment by drupol** [comment]
Question: Is the Nix indentation mechanism and function still necessary if you pass the resulting PHP code through PHP-CS-Fixer?

File: pkgs/pkgs-lib/formats.nix
Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix]: @drupol: Maybe I'm doing something wrong?
```
$ cat default.php
<?php

return ['attrs' => ['foo' => null],
'float' => 3.141000,
'int' => 10, 'str' => 'foo', 'hell' => ['attrs' => ['foo' => null], 'float' => 3.141000, 'int' => 10, 'str' => 'foo']];

$ php-cs-fixer fix --rules=@PSR12 default.php
PHP CS Fixer 3.51.0 Insomnia by Fabien Potencier, Dariusz Ruminski and contributors.
PHP runtime: 8.3.6
Loaded config default.
 1/1 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%


Fixed 0 of 1 files in 0.004 seconds, 14.000 MB memory used

$ cat default.php
<?php

return ['attrs' => ['foo' => null],
'float' => 3.141000,
'int' => 10, 'str' => 'foo', 'hell' => ['attrs' => ['foo' => null], 'float' => 3.141000, 'int' => 10, 'str' => 'foo']];
```

Comment (drupol) [reply pkgs/pkgs-lib/formats.nix]: You need to pass a directory, not a file.

Comment (drupol) [reply pkgs/pkgs-lib/formats.nix]: Passing a file will be soon available, ref: https://twitter.com/_Codito_/status/1789049130867650665

Comment (drupol) [reply pkgs/pkgs-lib/formats.nix]: Indeed, I'm also unable to get it working now. WTH ?!?!

**Comment by jtojnar** [comment]
> Question: Is the Nix indentation mechanism and function still necessary if you pass the resulting PHP code through PHP-CS-Fixer?

If we decide including PHP-CS-Fixer in the closure is fine, then Nix
indentation mechanism is pointless.
If we decide that the former is too heavy, then doing basic formatting
ourselves would be preferred.

**Comment by thenhnn** [comment]
> Question: Is the Nix indentation mechanism and function still necessary if you pass the resulting PHP code through PHP-CS-Fixer?
> If we decide including PHP-CS-Fixer in the closure is fine, then Nix indentation mechanism is pointless. If we decide that the former is too heavy, then doing basic formatting ourselves would be preferred.

I'm a bit confused, why format auto-generated file at all?

**Comment by jtojnar** [comment]
To make it easier for humans to read when debugging.

File: pkgs/pkgs-lib/formats.nix
Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix]: I'm afraid to write custom indentation code because I'm not a PHP programmer. Is there any other formatters for PHP?

Comment (drupol) [reply pkgs/pkgs-lib/formats.nix]: no.

Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix]: ugh... i tried php_codesniffer but it does nothing too.. php is cursed.

**Comment by jtojnar** [comment]
Formatting is just nice to have, IMO, so no need to block on that.

**Comment by thenhnn** [comment]
> Formatting is just nice to have, IMO, so no need to block on that.

Can we just omit it for now?

**Comment by thenhnn** [comment]
@jtojnar I think it's ready to merge.

File: pkgs/pkgs-lib/formats.nix
Comment (jtojnar) [reply pkgs/pkgs-lib/formats.nix]: Well, lot of small changes does add up too. But I do not feel strongly about this.

**Comment by jtojnar** [comment]
> Can we just omit it for now?

Yes, that is what I meant.



> I think it's ready to merge.

No other input from me, other than the minor preference for moving it to a separate file.

----

We should probably give @infinisil as a code owner a bit more time to react if they want.

**Comment by thenhnn** [comment]
> > Can we just omit it for now?
> 
> Yes, that is what I meant.
> 
> > I think it's ready to merge.
> 
> No other input from me, other than the minor preference for moving it to a separate file.
> 
> We should probably give @infinisil as a code owner a bit more time to react if they want.

I moved it to separate file, yeah, it's better :)

Got it, thank you!

File: pkgs/pkgs-lib/formats/php/default.nix
Comment (infinisil) [inline pkgs/pkgs-lib/formats/php/default.nix:55]: Why the assert here? If PHP supports mixed arrays with no indexed elements then this shouldn't be done

File: nixos/doc/manual/development/settings-options.section.md
Comment (infinisil) [inline nixos/doc/manual/development/settings-options.section.md:155]: It's a bit confusing on whether the `$` should be part of it. Consider either cleaning up the docs, or removing the need to specify the `$`

**Review by infinisil — CHANGES_REQUESTED** [review]
Not bad! Leaving some feedback :)

File: pkgs/pkgs-lib/formats/php/default.nix
Comment (Ma27) [reply pkgs/pkgs-lib/formats/php/default.nix:31]: (fwiw PHP supports trailing commas in lists. In fact, I know of coding styles that mandate those ;)

Comment (thenhnn) [reply pkgs/pkgs-lib/formats/php/default.nix:31]: Yeah, that's why there's no check

File: pkgs/pkgs-lib/formats/php/default.nix
Comment (thenhnn) [reply pkgs/pkgs-lib/formats/php/default.nix]: It's kinda bad, because if there's no indexed elements - why not use just set?

I replaced implementation to handle this case, not just throwing an error.

**Comment by thenhnn** [comment]
@jtojnar Can you approve changes? Or I missed something?

File: pkgs/pkgs-lib/formats/php/default.nix
Comment (infinisil) [reply pkgs/pkgs-lib/formats/php/default.nix]: This would be problematic if there's some third-party code making use of this function to just propagate some values, which would have to switch between a set and a mixed list depending on whether the list is empty. Not having an assertion avoids this

File: pkgs/pkgs-lib/tests/formats.nix
Comment (infinisil) [inline pkgs/pkgs-lib/tests/formats.nix:430]:
```suggestion
    format = formats.php { finalVariable = "config"; };
```

Though I don't see any test failure in CI 🤔

**Review by infinisil — APPROVED** [review]
Other than the above this looks fine to me!

File: pkgs/pkgs-lib/tests/formats.nix
Comment (thenhnn) [reply pkgs/pkgs-lib/tests/formats.nix:430]: oops D:

**Review by jtojnar — APPROVED** [review]

**Comment by thenhnn** [comment]
so, it's ready to merge, right?

**Comment by drupol** [comment]
Thank you!

**Pull request merged**
